File: Technical/3 ClassNotes/Java/src/practiceProblems/MetroTicketSimulation.java
package practiceProblems;

import java.util.HashMap;

public class MetroTicketSimulation {

    public static void main(String[] args) {

        int tickets[][] = {

                {101,1,5},
                {102,2,3},
                {103,1,4},
                {104,3,7},
                {105,2,2}

        };


        System.out.println(
                "Metro Ticket Records:");

        printMatrix(tickets);


        //Total Tickets
        int total =
                totalTickets(tickets);

        System.out.println(
                "\nTotal Tickets Sold: "
                + total);


        //Station with maximum tickets
        int station =
                maxTicketStation(tickets);

        System.out.println(
                "Station with Maximum Tickets: "
                + station);


        //String Compression
        String stations =
                "AAABBBCCDAA";

        System.out.println(
                "\nCompressed Stations: "
                + compress(stations));


        //Parsing
        String log =
                "101-Chennai-5";

        parseLog(log);

    }



    //Matrix Traversal
    static void printMatrix(
            int arr[][]) {

        for(int i=0;i<arr.length;i++) {

            for(int j=0;
                j<arr[i].length;
                j++) {

                System.out.print(
                        arr[i][j]
                        + " ");
            }

            System.out.println();
        }
    }



    //Total Tickets
    static int totalTickets(
            int arr[][]) {

        int total = 0;

        for(int i=0;i<arr.length;i++)
            total += arr[i][2];

        return total;
    }



    //Station with maximum tickets
    static int maxTicketStation(
            int arr[][]) {

        HashMap<Integer,Integer>
                map =
                new HashMap<>();

        for(int i=0;i<arr.length;i++) {

            int station =
                    arr[i][1];

            int count =
                    arr[i][2];

            map.put(
                    station,
                    map.getOrDefault(
                            station,
                            0)+count);
        }

        int maxStation = -1;
        int maxTickets = 0;

        for(Integer key :
                map.keySet()) {

            if(map.get(key)
                    > maxTickets) {

                maxTickets =
                        map.get(key);

                maxStation = key;
            }
        }

        return maxStation;
    }



    //String Compression
    static String compress(
            String str) {

        StringBuilder sb =
                new StringBuilder();

        int count = 1;

        for(int i=0;
            i<str.length()-1;
            i++) {

            if(str.charAt(i)
                    == str.charAt(i+1))

                count++;

            else {

                sb.append(
                        str.charAt(i));

                sb.append(count);

                count = 1;
            }
        }

        sb.append(
                str.charAt(
                        str.length()-1));

        sb.append(count);

        return sb.toString();
    }



    //Parsing
    static void parseLog(
            String log) {

        String parts[] =
                log.split("-");

        System.out.println(
                "\nParsed Log:");

        System.out.println(
                "Passenger ID: "
                + parts[0]);

        System.out.println(
                "City: "
                + parts[1]);

        System.out.println(
                "Tickets: "
                + parts[2]);
    }

}

File: Technical/3 ClassNotes/Python/practice_problems/metroTicketSimulation.py
class MetroTicketSimulation:


    # Matrix Traversal
    def print_matrix(self, arr):

        for row in arr:

            for val in row:
                print(val, end=" ")

            print()


    # Total Tickets
    def total_tickets(self, arr):

        total = 0

        for row in arr:
            total += row[2]

        return total


    # Station with Maximum Tickets
    def max_ticket_station(
            self,
            arr):

        freq = {}

        for row in arr:

            station = row[1]
            tickets = row[2]

            freq[station] = \
                freq.get(station,0) \
                + tickets

        max_station = max(
            freq,
            key=freq.get
        )

        return max_station


    # String Compression
    def compress(self, s):

        result = ""

        count = 1

        for i in range(len(s)-1):

            if s[i] == s[i+1]:

                count += 1

            else:

                result += s[i] \
                        + str(count)

                count = 1

        result += s[-1] + str(count)

        return result


    # Parsing
    def parse_log(self, log):

        parts = log.split("-")

        print("\nParsed Log:")

        print("Passenger ID:",
              parts[0])

        print("City:",
              parts[1])

        print("Tickets:",
              parts[2])



# Driver Code

obj = MetroTicketSimulation()

tickets = [

    [101,1,5],
    [102,2,3],
    [103,1,4],
    [104,3,7],
    [105,2,2]

]

print("Metro Ticket Records:")

obj.print_matrix(tickets)


total = obj.total_tickets(tickets)

print("\nTotal Tickets Sold:",
      total)


station = obj.max_ticket_station(
            tickets)

print(
    "Station with Maximum Tickets:",
    station
)


stations = "AAABBBCCDAA"

print(
    "\nCompressed Stations:",
    obj.compress(stations)
)


log = "101-Chennai-5"

obj.parse_log(log)